File: app/helpers/programs.py
import os
import asyncio
from app.sockets import sockets


programPath = os.path.join(os.path.dirname(__file__),'../../programs')

def listAll():
    fileList = [f for f in os.listdir(programPath) if os.path.isfile(os.path.join(programPath, f))]
    fileList = list(map(lambda f: f.replace('.json', ''), fileList))
    fileList.sort()
    return fileList

def getProgramJson(name):
    with open(os.path.join(programPath, name+'.json'), 'r') as myfile:
        return myfile.read()

def saveProgram(name, data):
    with open(os.path.join(programPath, name+'.json'), 'wb') as myfile:
        myfile.write(data)

def createProgram(name):
    saveProgram(name, '{"vars": [], "code": []}'.encode())

def deleteProgram(name):
    os.remove(os.path.join(programPath, name+'.json'))

def renameProgram(oldname, newname):
    os.rename(os.path.abspath(os.path.join(programPath, oldname+'.json')), os.path.abspath(os.path.join(programPath, newname+'.json')))


def runProgram(name):
    asyncio.ensure_future(_stream_subprocess(['python3', os.path.join(programPath, 'compiled', name+'.py')]))

async def _read_stream(stream, cb):
    chars = ''
    while True:
        char = await stream.read(1)
        if char:
            chars +=char.decode('utf-8')
            if chars.endswith('\n') or chars == '__RECEIVING_INPUT__':
                cb(chars)
                chars = ''
        else:
            break

def writeToProcess(line, process):
    process.stdin.write(str.encode(line+'\n'))

async def _stream_subprocess(cmd):
    process = await asyncio.create_subprocess_exec(*cmd, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE, stdin=asyncio.subprocess.PIPE)
    sockets.terminateListener = lambda: process.terminate()
    sockets.inputListener = lambda line: writeToProcess(line, process)
    await asyncio.wait([
        _read_stream(process.stdout, sockets.sendStdout),
        _read_stream(process.stderr, sockets.sendStderr)
    ])
    if process.returncode != None:
        process = None
        sockets.sendStdout('Program has finished.')
        sockets.inputListener = None
    else:
        await process.wait()
        process = None
        sockets.sendStdout('Program has finished.')
        sockets.inputListener = None

File: nodeapp/programs.js
const makePath = require('path').resolve
const fs = require('fs')
const spawn = require('child_process').spawn

const sockets = require('./sockets')
const compiler = require('./compiler')

const programPath = makePath(__dirname, '../programs')

module.exports = {
  listAll,
  getProgramJson,
  saveProgram,
  createProgram,
  deleteProgram,
  renameProgram,
  getCode,
  compileProgram,
  runProgram
}

function listAll (cb) {
  fs.readdir(programPath, (err, files) => {
    if (err)
      cb(err)
    else
      cb(null, files.filter(file => file != 'compiled').filter(file => file.substr(0, 1) != '.').map(file => file.replace('.json', '')).sort())
  })
}

function getProgramJson (name, cb) {
  const path = makePath(programPath, name+'.json')
  fs.readFile(path, 'utf8', cb)
}

function saveProgram(name, data, cb) {
  const path = makePath(programPath, name+'.json')
  fs.writeFile(path, data, cb)
}

function createProgram (name, cb) {
    saveProgram(name, '{"vars": [], "code": []}', cb)
}

function deleteProgram (name, cb) {
  const path = makePath(programPath, name+'.json')
  fs.unlink(path, cb)
}

function renameProgram (oldname, newname, cb) {
  var oldpath = makePath(programPath, oldname+'.json')
  var newpath = makePath(programPath, newname+'.json')
  fs.rename(oldpath, newpath, cb)
}

function getCode (name, cb) {
  const path = makePath(programPath, 'compiled', name+'.py')
  fs.readFile(path, 'utf8', cb)
}

function compileProgram (name, cb) {
  const inPath = makePath(programPath, name+'.json')
  const outPath = makePath(programPath, 'compiled', name+'.py')
  compiler.compileFile(inPath, outPath, cb)
}

function runProgram (name, cb) {
  compileProgram(name, err => {
    if (err) {
      cb(err)
      return
    }
    executePython(makePath(programPath, 'compiled', name.replace(/ /, '\\ ')+'.py'))
    cb(null)
  })
}

function executePython (path) {
  const process = spawn('python3 '+path, {shell: true, stdio: 'pipe'})
  process.on('close', () => {
    sockets.setStdinHandler(null)
    sockets.setTerminateHandler(null)
    sockets.sendStdout('Program has finished.')
  })
  process.on('error', () => {
    sockets.setStdinHandler(null)
    sockets.setTerminateHandler(null)
    sockets.sendStdout('Program has finished.')
  })
  sockets.setTerminateHandler(process.kill)
  sockets.setStdinHandler(data => {
    process.stdin.write(data+'\n')
  })
  process.stdout.on('data', data => {
    console.log('stdout')
    console.log(data.toString('utf8'))
    sockets.sendStdout(data.toString('utf8'))
  })
  process.stderr.on('data', data => {
    console.log('stderr')
    console.log(data.toString('utf8'))
  })
}